### ISSUE TITLE
DIS: Keywords for multi-threading capabilities

### ISSUE THREAD

**Issue by lithomas1** [pandas-dev/pandas#43313]
With the addition of the new pyarrow engine, we now have the option to use multiple threads to read a CSV file. (This is also controllable through the ``pyarrow.set_cpu_count`` option). 

Should we expose a keyword(such as ``num_threads`` maybe) to the user as a keyword, or just add an example in the docs(for this case, redirecting to ``pyarrow.set_cpu_count``? In the case of ``read_csv``, this keyword would probably only apply to the ``pyarrow`` engines, however it is worth noting that we have had multiple feature requests for parallel CSV reading (e.g. #37955), and it is probably worth it to be configure the number of threads used if we offer multithreading. 

Personally, I would prefer having a keyword, as if we decide to add more I/O engines with multithreading capabilities, it would be more convenient to be able to control this option through a keyword.

cc @pandas-dev/pandas-core

**Comment by mroeschke** [comment]
What is done for numba capabilities (`engine="numba"`) and is the most flexible IMO is to pair `engine` with a  `engine_kwargs` keywords which accepts a `dict` with engine configurations e.g. `{"set_cpu_count": 2}`

This way the pandas API isn't tied to engine specific constructs while catering to whatever capabilities each engine supports.

**Comment by jreback** [comment]
we pass thru n_threads
in the parquet reader 

so wouldn't object to a matching keyword or @mroeschke suggestion

**Comment by lithomas1** [comment]
Personally, I think it might be better to just use a standard keyword argument, since it could be confusing if this changes between engines. I think this option would probably be popular enough to warrant this. 

Having a standard keyword to control multi-threading could also pave the way for a global option to config this for everything.

(btw, n_threads doesn't exist in pyarrow anymore, you have to use ``pyarrow.set_cpu_count``)

WDYT?

**Comment by brianpenghe** [comment]
Has this been solved? Reading a large csv is too slow with only one thread!

**Comment by phofl** [comment]
It does not seem so.

**Comment by jreback** [comment]
you can multi threaded reading with the engine='pyarrow' (in 1,4)

**Comment by jbrockmendel** [comment]
Out of curiosity, in what scenario with multi-threading available would you not want to crank it up to 11?  I guess debugging?

**Comment by bashtage** [comment]
> Out of curiosity, in what scenario with multi-threading available would you not want to crank it up to 11? I guess debugging?

When you use multi-process you almost always want to limit or eliminate multi-threading.  For example, when using linear algebra on a spark cluster, failing to set `MKL_NUM_THREADS=1` or `OMP_NUM_THREADS=1` can lead to a situation where you end up with `ncore * ncore` threads competing for ncore resources, and basically crippling congestion.

**Comment by rhshadrach** [comment]
> Out of curiosity, in what scenario with multi-threading available would you not want to crank it up to 11? I guess debugging?

I'll also add multi-user servers where you want to be (somewhat) nice to other processes.

**Comment by jorisvandenbossche** [comment]
In general when you have multiple libraries or functions each using multi-threading with their own thread pool (nested parallelism), you can easily get what is called "oversubscription". 

Typical example is scikit-learn where you could do several runs of a model with varying parameters (eg for parameter optimization) in parallel, but then the individual model algorithm (eg using MKL or BLAS) might again run things in parallel . They developed a package to better be able control the number of threads that varying libraries use (https://github.com/joblib/threadpoolctl) to cope with such kind of issues.

**Comment by jbrockmendel** [comment]
Thanks for those explanations of use cases for disabling/limiting multi-threading.  They all seem like cases where you'd want to do it across the board at a process level, so seem like a use case for pd.set_option as opposed to fine-grained keywords.

**Comment by WillAyd** [comment]
Out of curiosity I wanted to see what `callgrind` thought about our read_csv timings, using the 5GB csv file of Oct data taken from here:

https://www.kaggle.com/datasets/mkechinov/ecommerce-behavior-data-from-multi-category-store/discussion

Here are the results:

![read_csv](https://user-images.githubusercontent.com/609873/219271385-71403258-41b1-4928-abae-b28d5dd8c046.png)

I am by no means an expert on the topic, but my initial thought in reading these results is that I/O itself is not much of a bottleneck, and by extension I'm not sure adding threads to the mix would buy us a lot. The big bottlenecks appear to be the `tokenize_bytes` and `convert_with_dtype` functions (the latter especially with strings).

**Comment by jorisvandenbossche** [comment]
> I/O itself is not much of a bottleneck, and by extension I'm not sure adding threads to the mix would buy us a lot

But also the parsing/tokenizer step can in theory be parallelized?

**Comment by Dr-Irv** [comment]
> But also the parsing/tokenizer step can in theory be parallelized?

Beware of the GIL!  I was thinking  you could do something by chopping up the file horizontally (e.g., if you have 1M rows, split it into 4 pieces of 250K rows each), process the splits in parallel into 4 DataFrames, then `concat` all of them.

But I think (although I'm not sure) that you might not get speedups because the different threads would get locks due to the GIL.  Even if the GIL isn't involved, if they are sharing a `malloc()` operation way down under the hood, there will be a lock there that could end up constraining any speedup.  I ran into this many years ago when writing some code in C that parallelized an algorithm.

**Comment by WillAyd** [comment]
I think it can be parallelized but I also think we are mixing threads and processes. Assuming the tokenizer is CPU bound (which I believe it is - don't see any IO therein) adding threads on the same process isn't going to help the processing time, only hurt it. 

Multiprocessing would seem a better fit, but the disadvantage I think we have is a lack of well defined IPC like Arrow has. The Python multiprocessing library  has an Array object that might help

**Comment by jbrockmendel** [comment]
This is worth reviving with free-threading on the horizon.  In read_csv we can potentially parallelize by splitting the file into chunks as Irv mentioned, and by parallelizing calls to `_convert_column_data`.  xref #61825.

The other obvious target is BlockManager.apply which sends a ton of operations through block-by-block iteration.

I expect that splitting the file will be the easiest perf improvement (for large files at least), but will necessitate a parameter like pyarrow's `newlines_in_values`.  (I kind of like pyarrow's ParseOptions pattern).  

Adding a keyword to all the methods that go through BM.apply would be really heavy-handed.  Maybe a global option would be nicer? (scrolling up I see past-me thought that too)

**Comment by rhshadrach** [comment]
+1 on the global option, but we will need to be careful about stacking calls to multithreaded methods. If I call something that kicks off 32 threads and each of those calls a method that again spawns 32 threads, then the user will encounter a lot of thrashing. 

Also if we're not reusing a global thread pool, then just spawning the threads can have significant overhead I think. We could see some real slow downs on small data. If we can't use a thread pool and have a global option, do we then need to only use multithreading when the data size justifies it?

**Comment by jbrockmendel** [comment]
Discussed this on the dev call before last (maybe one before that?).  Agreed that if a user is using a free-threading build, that is a good indication they want to opt-in to parallelism.  We can let them opt back out with a global option.
